Define net.sock attributes and clarify logical net.peer|host.name attributes

lmolkova opened the pull request:
Fixes #2607

## Changes

1. Introduces socket-level attributes: 
   - BREAKING: `net.peer.ip` ->  `net.sock.peer.addr`, `net.host.ip` ->  `net.sock.host.addr`
   - BREAKING: for pure socket-level instrumentation, `net.peer.name` is renamed to `net.sock.peer.addr`. For other instrumentations, `net.peer.name` semantics didn't change.
3. Clarifies `net.peer|host.name`  and port attributes as logical ones representing final destination.
4. Clarifies socket-level proxy scenario

Related issues #2469

lmolkova commented:
@Oberon00 I made the first stab at `net.sock` attributes. 

One thing I didn't figure out yet is the relationship between `net.transport` and `net.sock.family`. 

Transport currently is a closed set, address families are not (and different sets on win and linux).
Transport defines TCP and UDP, address family differentiates IPv4 and IPv6.
Transport has `pipe`  and it's not a socket.

I'd propose the following on '`net.transport`:
- remove or deprecate `net.transport` `ip` and `unix` values
- add `socket` value

Then `net.sock.family` can be used to understand if it's an IP-based protocol, Unix, or perhaps Bluetooth when `net.transport = socket`. Or if addr family is `AF_INET6`,  `net.transport` explains if it tcp or udp. 

LMK what you think.

Oberon00 commented:
I didn't have time to think about the net.transport question yet, but I noticed one thing:

There is no replacement for net.peer.name yet, and also the definition in the "net.*.name" section has not changed yet to address the logical peer change:

> For net.peer.name, this should be the name of logical remote destination that was used to look up the IP address that was connected to (i.e., matching net.peer.ip if that one is set)

I also think that net.peer.ip does not make sense anymore, there should probably only be net.sock.peer.ip.

lmolkova commented:
> I also think that net.peer.ip does not make sense anymore, there should probably only be net.sock.peer.addr.

I think we wanted to use it instead of `http.client_ip`, but thinking more about it, perhaps reverse proxy and `x-forwarded-for` and friends are so HTTP-specific that `http.client_ip` would make more sense than generic `net.peer.ip`.

lmolkova commented:
@Oberon00 I think I addressed your feedback. I'm not sure if renaming `net.peer.name` is necessary - even though we work on experimental spec, we should minimize breaking changes. Currently, the breaking change is minimal (`net.peer.ip` -> `net.sock.peer.addr` and `net.host.ip` ->  `net.sock.host.addr`) and spec is clear on socket vs logical attributes plus proxy scenarios,

Oberon00 commented:
I'm not talking about renaming of net.peer.name, but about adding an additional attribute, e.g. net.sock.peer.name that would then correspond to net.sock.peer.addr (net.peer.name does not necessarily do that anymore).

lmolkova commented:
@Oberon00  what would `net.sock.peer.name` mean - DNS name corresponding for `net.sock.peer.addr` ? Like proxy host name if one's available?

Oberon00 commented:
> @Oberon00 what would net.sock.peer.name mean - DNS name of net.sock.peer.addr ? Like proxy host name if one's available?

Exactly. This is how I understood the net.peer.ip/name attributes originally. net.sock.peer.name would then be mostly IP-specific, just like net.peer.ip currently is.

As I understand it, this PR will cause the following "reshuffling":

* net.peer.name: Logical peer name, but not of the peer service (which is peer.service) but of something that could theoretically be "connected" to, typically a DNS-resolvable hostname  (even if just from some private network), but might be something else. Depending on the address family used, a low-level 2non-logical" net.peer.name becomes net.peer.sock.addr or net.peer.sock.name.
* net.peer.ip -> net.peer.sock.addr, always, and only available for low-level peer, not for logical one if different.

lmolkova commented:
@Oberon00 thank you for the feedback! I believe I addressed it and this PR now is ready for review.

Oberon00 commented on CHANGELOG.md:
```diff
@@ -23,6 +23,9 @@ release.
   ([#2602](https://github.com/open-telemetry/opentelemetry-specification/pull/2602)).
 - Adopt attribute requirement levels in semantic conventions
   ([#2594](https://github.com/open-telemetry/opentelemetry-specification/pull/2594))
+- BREAKING: rename `net.peer.ip` to `net.sock.peer.addr` and `net.host.ip` to `net.sock.host.addr`.
```
```suggestion
- BREAKING: rename `net.peer.ip`, and `net.peer.name` in some usages, to `net.sock.peer.addr` and `net.host.ip` to `net.sock.host.addr`.
```

Oberon00 commented on semantic_conventions/trace/general.yaml:
```diff
@@ -48,31 +42,67 @@ groups:
+          allow_custom_values: true
+          members:
+            - id: inet
+              value: 'AF_INET'
```
```suggestion
              value: 'inet'
```

If we want to reference the Unix C define name, we can add it to brief.

Oberon00 commented on semantic_conventions/trace/general.yaml:
```diff
@@ -48,31 +42,67 @@ groups:
+              value: 'AF_INET'
+              brief: "IPv4 address"
+            - id: inet6
+              value: 'AF_INET6'
```
```suggestion
              value: 'inet6'
```

Oberon00 commented on semantic_conventions/trace/general.yaml:
```diff
@@ -48,31 +42,67 @@ groups:
+              value: 'AF_INET6'
+              brief: "IPv6 address"
+            - id: unix
+              value: 'AF_UNIX'
```
```suggestion
              value: 'unix'
```

Oberon00 commented on semantic_conventions/trace/general.yaml:
```diff
@@ -48,31 +42,67 @@ groups:
+              value: 'AF_UNIX'
+              brief: "Unix domain socket path/"
+        requirement_level:
+          conditionally_required: if different than `AF_INET` and if any of `net.sock.peer.addr` or `net.sock.host.addr` are set
```
```suggestion
          conditionally_required: if different than `inet` and if any of `net.sock.peer.addr` or `net.sock.host.addr` are set
```

lmolkova commented:
@Oberon00 I believe I resolved your feedback, can you please take another look?

Oberon00 reviewed (approved):
I wonder if we also need net.host.sock.*?

lmolkova commented:
@Oberon00 

> I wonder if we also need net.sock.host.*?

I defined `net.sock.host.addr` and  `net.sock.host.port` and briefly described them [here](https://github.com/open-telemetry/opentelemetry-specification/blob/39ec7eeb6f6804dd8294864d0873171d5eb73994/specification/trace/semantic_conventions/span-general.md#host) Do you think something important is missing?

Oberon00 commented on semantic_conventions/trace/general.yaml:
```diff
@@ -6,18 +6,12 @@ groups:
     attributes:
```
I think at this point it would make sense to split net.* into a separate file (also the markdown)

lmolkova replied:
Yes, it makes sense, Can I do it in a separate PR? I want to unblock the HTTP - it's been open for a while.

Oberon00 commented on semantic_conventions/trace/http.yaml:
```diff
@@ -114,10 +114,17 @@ groups:
+      - ref: net.sock.peer.addr
+      - ref: net.sock.peer.port
+        requirement_level:
+          recommended: If different than `net.peer.port` and if `net.sock.peer.addr` is set.
```
I think this (and similar ones) should go to the primary definition.

lmolkova replied:
Apparently, requirement level and msg references don't work - fixing it here https://github.com/open-telemetry/build-tools/pull/102

Oberon00 replied:
IIRC this was originally intentional, but I don't see any benefit of that design now.

Oberon00 commented on semantic_conventions/trace/general.yaml:
```diff
@@ -48,31 +42,67 @@ groups:
+              value: 'unix'
+              brief: "Unix domain socket path/"
+        requirement_level:
+          conditionally_required: if different than `inet` and if any of `net.sock.peer.addr` or `net.sock.host.addr` are set
```
I think this requirement is incompatible with the schema. Previously, we did not differentiate inet and inet6 in net.peer.ip. Now if the schema renames this to net.peer.sock.addr, and it contains an IPv6, then this would result in bad data.

I don't know how to handle this nicely, but some not so nice alternatives:
- Only have `inet`, rely on addr format to distinguish. If no addr is set, then we don't know.
- Add inet_any.
- Specify that if not set, inet6 vs inet must be determined by looking at the address

Maybe the last one would be quite alright:
```suggestion
          conditionally_required: if neither `inet` nor `inet6` and if any of `net.sock.peer.addr` or `net.sock.host.addr` are set.
```

lmolkova commented:
@carlosalberto can you please take a look?

lmolkova commented:
@open-telemetry/specs-approvers can someone please take a look? It unblocks https://github.com/open-telemetry/opentelemetry-specification/pull/2469

trask commented on specification/metrics/semantic_conventions/http-metrics.md:
```diff
@@ -52,7 +52,7 @@ and whether they should be on server, client, or both types of HTTP metric event
 | `net.peer.name`    | `client`            | see [1] in [attribute alternatives](#attribute-alternatives) | See [general network connection attributes](../../trace/semantic_conventions/span-general.md#general-network-connection-attributes)                                                                                 |
 | `net.peer.port`    | `client`            | see [1] in [attribute alternatives](#attribute-alternatives) | See [general network connection attributes](../../trace/semantic_conventions/span-general.md#general-network-connection-attributes)                                                                                 |
-| `net.peer.ip`      | `client`            | see [1] in [attribute alternatives](#attribute-alternatives) | See [general network connection attributes](../../trace/semantic_conventions/span-general.md#general-network-connection-attributes)                                                                                 |
+| `net.sock.peer.addr`      | `client`            | see [1] in [attribute alternatives](#attribute-alternatives) | See [general network connection attributes](../../trace/semantic_conventions/span-general.md#general-network-connection-attributes)                                                                                 |
```
attribute alternatives is referenced here, but ip/addr was deleted from attribute alternatives below

trask commented on specification/trace/semantic_conventions/database.md:
```diff
@@ -42,21 +42,28 @@ Some database systems may allow a connection to switch to a different `db.user`,
-| [`net.peer.name`](span-general.md) | string | Remote hostname or similar, see note below. [1] | `example.com` | Conditionally Required: See alternative attributes below. |
-| [`net.peer.port`](span-general.md) | int | Remote port number. | `80`; `8080`; `443` | Conditionally Required: [2] |
-| [`net.transport`](span-general.md) | string | Transport protocol used. See note below. | `ip_tcp` | Conditionally Required: [3] |
+| [`net.peer.name`](span-general.md) | string | Name of the database host. [1] | `example.com` | Conditionally Required: See alternative attributes below. |
+| [`net.peer.port`](span-general.md) | int | Logical remote port number | `80`; `8080`; `443` | Conditionally Required: [2] |
```
are these only for client now? if so, maybe we should add column to the table to identify client and/or server similar to http metrics attributes table:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#attributes

trask commented on semantic_conventions/trace/general.yaml:
```diff
@@ -45,34 +39,72 @@ groups:
+        examples: '192.168.0.1'
+      - id: sock.host.port
+        type: int
+        brief: 'Local socket peer port (if defined for the address family).'
```
```suggestion
        brief: 'Local socket port number.'
        requirement_level:
          recommended: If defined for the address family and if different than `net.host.port` and if `net.sock.host.addr` is set.
```

trask reviewed (approved):
fyi we've prototyped this change in Java instrumentation: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6268

lmolkova commented:
@trask thanks a lot for the feedback and the prototype!

@carlosalberto, can you please take another look?  Thanks!

jmacd commented:
@open-telemetry/specs-approvers Last call for feedback.

tigrannajaryan commented on specification/trace/semantic_conventions/span-general.md:
```diff
@@ -37,12 +43,17 @@ the `net.peer.*` properties of a client are equal to the `net.host.*` properties
-| `net.host.port` | int | Like `net.peer.port` but for the host port. | `35555` | Recommended |
-| `net.host.name` | string | Local hostname or similar, see note below. | `localhost` | Recommended |
+| `net.sock.peer.name` | string | Remote socket peer name. | `proxy.example.com` | Recommended: [2] |
+| `net.sock.peer.addr` | string | Remote socket peer address (IPv4 or IPv6 for internet protocols, path for local communication,
```
The formatting of this line is broken. Check the rendered markdown.

lmolkova replied:
great catch, fixed!

lmolkova commented on schemas/1.12.0:
```diff
@@ -2,6 +2,12 @@ file_format: 1.0.0
+        - rename_attributes:
+            attribute_map:
+              net.peer.ip: net.sock.peer.addr
+              net.host.ip: net.sock.host.addr
```
> I assume the reason why we don't have a line for net.peer.name here is because its semantics didn't change for most case but it was renamed for specific entities only.

Yes and because we didn't have socket-level semantics clearly defined before.

I double-checked the document status on each document changed and all of the conventions are experimental (to be fair, we don't have non-experimental semantic conventions yet).
